kernel/mempool.c

```diff
@@ -46,9 +62,24 @@ void k_free(void *ptr)
 K_HEAP_DEFINE(_system_heap, CONFIG_HEAP_MEM_POOL_SIZE);
 #define _SYSTEM_HEAP (&_system_heap)
 
-void *k_malloc(size_t size)
+void *k_aligned_alloc(size_t align, size_t size)
 {
-	return z_heap_malloc(_SYSTEM_HEAP, size);
+	/* align must be a power of two */
+	if (popcount(align) != 1) {
+		return NULL;
+	}
+
+	/* align must be multiple of sizeof(void *) */
```

Comment:
same comment twice, but these are different checks anyway

Where does the multiple of sizeof(void *) requirement come from anyway? Does the underlying sys_heap require it? If so, shouldn't the checks be moved there.

At minimum, these should be assertions with __ASSERT and not unconditonally checked

Reply:
> same comment twice, but these are different checks anyway
Both checks are actually to fulfil the same requirement (with the convention that an integral multiple is >= 1) and for that reason the comment is duplicated. I'll remove second comment though and refine the first.

> Where does the multiple of sizeof(void *) requirement come from anyway? Does the underlying sys_heap require it? If so, shouldn't the checks be moved there.

The usual place - [standards](https://linux.die.net/man/3/posix_memalign). Supporting arbitrary alignments (e.g. 1, 2, 3, 7) is not usually needed.

> At minimum, these should be assertions with __ASSERT and not unconditonally checked

Sure, I can do that. Popcount is eventually checked in `sys_heap_alloc_aligned()`. I will need to remove some test points as well to avoid `__ASSERT` triggering. Also, it makes the implementation a bit less like the semantic equivalent I was aiming for.